mitodl/open-discussions
# 4228 keycloak integration login register password reset ux

#### Pre-Flight checklist

#### What are the relevant tickets?
https://github.com/mitodl/open-discussions/issues/4228

#### What's this PR do?
When the environment variable `FEATURE_KEYCLOAK_ENABLED` is set to True, all password, registration, and login functionality should be handled by Keycloak.  The default value for `FEATURE_KEYCLOAK_ENABLED` is False.

When `FEATURE_KEYCLOAK_ENABLED` is `True`:
- URLs no longer needed should be disabled.
- Login and Sign up UI and URLs should initiate the Keycloak login workflow.
- Password reset URL's and UI should initiate the Keycloak password change workflow.
- Logging out of Open Discussions should terminate the user's Keycloak session.

When `FEATURE_KEYCLOAK_ENABLED` is `False`: 
- All functionality added by this PR should be disabled.

3 environment variables have been added and are required:

- `FEATURE_KEYCLOAK_ENABLED`
- `KEYCLOAK_BASE_URL`
- `KEYCLOAK_REALM_NAME`

## How should this be manually tested?
These instructions assume that you have Keycloak in the QA environment configured with your local instance of Open Discussions.

- Add `KEYCLOAK_BASE_URL=https://sso-qa.odl.mit.edu` to your .env file.
- Add `KEYCLOAK_REALM_NAME=olapps` to your .env file.
### Testing URLs
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`.
2. The following URL endpoints should redirect to Keycloak's login screen: http://od.odl.local:8063/login/, http://od.odl.local:8063/signup/.
3. The following URL endpoints should not respond: http://od.odl.local:8063/password_reset/, http://od.odl.local:8063/signup/confirm/, http://od.odl.local:8063/password_reset/confirm/, http://od.odl.local:8063/api/v0/login/email/, http://od.odl.local:8063/api/v0/login/password/, http://od.odl.local:8063/api/v0/register/email/, http://od.odl.local:8063/api/v0/register/confirm/, http://od.odl.local:8063/api/v0/register/details/, http://od.odl.local:8063/api/v0/password_reset/, http://od.odl.local:8063/api/v0/password_reset/confirm/, http://od.odl.local:8063/api/v0/set_password/
4. The following API endpoints should be accessible using http://od.odl.local:8063/api/v0/auth/<EMAIL> and http://od.odl.local:8063/api/v0/update_password_request/.

### Testing Login UI
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`.
2. The following UI elements should redirect to Keycloak's login screen: 
- Login option from top right menu.  
- Login option shown when clicking "Create a post" from the home screen.
3. Login using one of the options from step 2. 

### Testing Password Change
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`.
2. Follow the steps in "Testing Login UI".
3. Navigate to http://od.odl.local:8063/settings/password which should redirect you to the Keycloak password change screen.  There should also be a link at the top-right corner of the page which allows you to navigate back to Open Discussions.

### Testing Logout
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`
2. In a browser tab, follow the steps in "Testing Login UI".
4. In the Open discussions browser tab, click "Sign Out" from the top-right menu of Open Discussions or visit http://od.odl.local:8063/logout.
5. You should be redirected to a Keycloak page asking you to confirm your logout.  Confirm the logout.
6. Visit http://od.odl.local:8063/ and confirm you are logged out of Open Discussions.
7. Visit https://sso-qa.odl.mit.edu/realms/olapps/account/ and confirm you are logged out of Keycloak.

### Testing the environment variable as `False`.
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=False`.
9. Restart the application.
10. All URL endpoints and UI related to passwords, registration, and login functionality should perform as they were before this PR.
11. The following API endpoints should NOT be accessible using http://od.odl.local:8063/api/v0/auth/<EMAIL> and http://od.odl.local:8063/api/v0/update_password_request/.

### Testing default environment variable value
1. In the `.env` file, remove `FEATURE_KEYCLOAK_ENABLED=False`.
6. Restart the application.
8. All URL endpoints and UI related to passwords, registration, and login functionality should perform as they were before this PR.
9. The following API endpoints should NOT be accessible using http://od.odl.local:8063/api/v0/auth/<EMAIL> and http://od.odl.local:8063/api/v0/update_password_request/.
